src/algebra/category/Module/change_of_rings.lean

```diff
@@ -0,0 +1,84 @@
+/-
+Copyright (c) 2022 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+import algebra.category.Module.basic
+import algebra.category.Algebra.basic
+import ring_theory.tensor_product
+
+/-!
+# Change Of Rings
+
+## Main definitions
+
+* `category_theory.Module.restrict_scalars`: given rings `R, S` and a ring homomorphism `R ⟶ S`,
+  then `restrict_scalars : Module S ⥤ Module R` is defined by `M ↦ M` where `M : S-module` is seen
+  as `R-module` by `r • m := f r • m` and `S`-linear map `l : M ⟶ M'` is `R`-linear as well.
+
+## List of notations
+Let `R, S` be rings and `f : R →+* S`
+* if `M` is an `S`-module, `r : R` and `m : M` then notation `r r•[f] m` means `R`-scalar action on
+  `M` defined by `f r • m`.
+-/
+
+
+namespace category_theory.Module
+
+universes v u₁ u₂
+
+namespace restrict_scalars
+
+variables {R : Type u₁} {S : Type u₂} [ring R] [ring S] (f : R →+* S)
+
+section unbundled
+
+variables (M : Type v) [add_comm_monoid M] [module S M]
+
+/-- The `R`-scalar multiplication on `S`-module M defined by `r • m := f r • m` -/
+protected def has_smul : has_smul R M :=
+(module.comp_hom M f).to_has_smul
+
+localized "notation r ` r•[` f `] ` :=
+  @@has_smul.smul (restrict_scalars.has_smul f _) r" in change_of_rings
+
+end unbundled
+
+open_locale change_of_rings
+
+variable (M : Module.{v} S)
+
+/-- Any `S`-module M is also an `R`-module via a ring homomorphism `f : R ⟶ S` by defining
+    `r • m := f r • m` (`module.comp_hom`). This is called restriction of scalars. -/
+def obj' : Module R :=
+{ carrier := M,
+  is_add_comm_group := infer_instance,
+  is_module := module.comp_hom M f }
+
+
+@[simp] lemma smul_def (r : R) (m : M) : r r•[f] m = f r • m := rfl
+
+/--
+Given an `S`-linear map `g : M → M'` between `S`-modules, `g` is also `R`-linear between `M` and
+`M'` by means of restriction of scalars.
+-/
+@[simps] def map' {M M' : Module.{v} S} (g : M ⟶ M') :
+  obj' f M ⟶ obj' f M' :=
+{ map_smul' := λ r (x : M), by simp,
```

Comment:
If I replace the `by simp` by an empty `begin end` and look at the goal view, I see
```lean
g.to_fun (r • x) = ⇑(ring_hom.id R) r • g.to_fun x
```
I would have hoped that instead of `r • x` it said `r r•[f] x`. Did this work before you unbundled the notation? I guess the point of the notation is more to use it in statements rather than have Lean show it in the goal view, but it still would be nice if this worked.

Reply:
The goal is the same no matter if notation is bundled or not

Reply:
The weird thing is that even if I `change g (r r•[f] _) = _,`, the goal still doesn't display `r r•[f] x`, instead it still says `g.to_fun (r • x) = ...`

Reply:
Would you mind extracting this into a MWE and asking on Zulip?

Reply:
> I think the actual problem here is that lean won't ever pretty print notation with a @ in its definition; not that it's choosing the existing notation over your notation.

Should we just delete this notation

Reply:
I have deleted the notation